gcloud-java-bigquery/src/main/java/com/google/cloud/bigquery/Job.java

```diff
@@ -144,14 +163,51 @@ public boolean isDone() {
   }
 
   /**
+   * Waits until this job completes its execution, either failing or succeeding. If the job does not
+   * exist, this method returns without executing any method of the provided callback. If the job
+   * completed successfully the {@link CompletionCallback#success(Job)} method is called. If the job
+   * completed with errors the {@link CompletionCallback#error(BigQueryError, List)} method is
+   * called.
+   * <pre> {@code
+   * job.whenDone(new CompletionCallback() {
+   *   void success(Job job) {
+   *     // completed successfully
+   *   }
+   *
+   *   void error(BigQueryError error, List<BigQueryError> executionErrors) {
+   *     // handle error
+   *   }
+   * });}</pre>
+   *
+   * @throws BigQueryException upon failure
+   * @throws InterruptedException if the current thread gets interrupted while waiting for the job
+   *     to complete
+   */
+  public void whenDone(CompletionCallback callback) throws InterruptedException {
```

Comment:
I am not too excited about this method even though it does reduce some boilerplate code.
I think it would have been much more useful if it was not blocking and we could return a "listenable future".

My issues with it are:

1) It is not clear that this method is blocking.Though Java doc says "Waits until..." I would be more explicit saying "this method will block until..." and also change the name of the method to something like `waitFor`.

2)  Existing the method without a callback *and* without indicating that a callback was not called is error prone (e.g. imagine callbacks were used to toggle some latch/semaphore). I think we should either make the method  return a boolean or add another explicit callback for not found.

3) There is actually another exit point in this case and that could be that the RPC for `isDone` or `reload` failed, which similar to (2) makes the callback usage more complex. A solution for that would be to never fail 
calling this method and to indicate any failures (in a distinct way) in the callback.

4) `500L` sampling rate may be too course or to grain depending on the job and the client needs. Maybe we could provide a var-arg way to tune it.